### PR TITLE
Support for AddCaptioning

### PR DESCRIPTION
**Pull request by amhed**
Hey guys, 

We've been using the API and found that the `add_captioning` method was missing from the wrapper, so we added it according to https://docs.brightcove.com/en/video-cloud/media/references/reference.html#Video_Write

Thanks again!

### CODE REVIEW

File: source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs [hunk 1]
```diff
@@ -430,18 +430,86 @@ private BrightcoveImage DoAddImage(BrightcoveImage image, FileUploadInfo fileUpl
...
+
+        #region AddCaptioning
+        public BrightcoveCaptioning AddCaptioning(string captionFileUrl, long videoId, string filename)
+        {
+            return DoAddCaptioningWithExternalUrl(captionFileUrl, videoId, null, filename);
+        }
+
+        public BrightcoveCaptioning AddCaptioning(string captionFileUrl, string videoReferenceId, string filename)
+        {
+            return DoAddCaptioningWithExternalUrl(captionFileUrl, -1, videoReferenceId, filename);
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs]: Do we need both of these methods? I don't think we gonna use `videoReferenceId`

File: source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs [hunk 2]
```diff
@@ -430,18 +430,86 @@ private BrightcoveImage DoAddImage(BrightcoveImage image, FileUploadInfo fileUpl
...
+                DisplayName = filename,
+                Url = captionFileUrl
+            };
+
+            BrightcoveParamCollection parms = CreateWriteParamCollection("add_captioning",
+                                                                         methodParams => {
+                                                                             methodParams.Add(propName, propValue);
+                                                                             methodParams.Add("caption_source", captionSource);
+                                                                             methodParams.Add("filename", filename);
+                                                                         });
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs]: my eyes are bleeding :)

File: source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs (hunk 1 above)
Comment (amhed) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs]: @kish9nine told me we will only be using videoId, but I added them both to comply with existing conventions

File: source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs [hunk 3]
```diff
@@ -430,18 +430,86 @@ private BrightcoveImage DoAddImage(BrightcoveImage image, FileUploadInfo fileUpl
...
+        private BrightcoveCaptioning DoAddCaptioningWithFileUpload(FileUploadInfo captionFileUploadInfo, long videoId, string videoReferenceid, string filename)
+	    {
+            string propName;
+            object propValue;
+            GetIdValuesForUpload(videoId, videoReferenceid, "video_id", "video_reference_id", out propName, out propValue);
+
+            var captionSource = new BrightcoveCaptionSource {
+                Complete = true,
+                IsRemote = false,
+                DisplayName = filename,
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs]: I doubt we want to use filename as caption track name. Usually it's just 'English'

File: source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs [hunk 4]
```diff
@@ -430,18 +430,86 @@ private BrightcoveImage DoAddImage(BrightcoveImage image, FileUploadInfo fileUpl
...
+        }
+
+        public BrightcoveCaptioning AddCaptioning(FileUploadInfo captionFileUploadInfo, long videoId, string filename)
+        {
+            return DoAddCaptioningWithFileUpload(captionFileUploadInfo, videoId, null, filename);
+        }
+
+        public BrightcoveCaptioning AddCaptioning(FileUploadInfo captionFileUploadInfo, string videoReferenceId, string filename)
+        {
+            return DoAddCaptioningWithFileUpload(captionFileUploadInfo, -1, videoReferenceId, filename);
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs]: I guess the pattern of usage would be uploading to Brightcove and not hosting it (using isRemote), right?

File: source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    /// </summary>
+    /// <remarks>
+    /// See https://docs.brightcove.com/en/video-cloud/media/references/reference.html#CaptionSource
+    /// </remarks>
+    public class BrightcoveCaptionSource : BrightcoveItem, IJavaScriptConvertable
+    {
+        /// <summary>
+        /// A Boolean indicating whether a CaptionSource is usable.
+        /// </summary>
+        public bool Complete { get; set; }
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs]: Their docs explanation sucks. My understanding is that equals 'is published' or 'not a draft'. Just speaking out loud

Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs]: Docs say it's read-only, so I guess we don't need to set it in the methods above.
Same applies to `isRemote`. It returns true for videos having `url`.

File: source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        public bool IsRemote { get; set; }
+
+        /// <summary>
+        /// The complete path to the file.
+        /// </summary>
+        public string Url { get; set; }
+
+        public IDictionary<string, object> Serialize(JavaScriptSerializer serializer)
+        {
+            IDictionary<string, object> serialized = new Dictionary<string, object>();
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs]: var?

File: source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+            serialized["complete"] = Complete;
+            serialized["displayName"] = DisplayName;
+            serialized["id"] = Id ?? 0;
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs]: Same concern about read-only fields. No need to set them, assuming Searilize is used to send data.

File: source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            serialized["isRemote"] = IsRemote;
+            serialized["url"] = Url;
+
+            return serialized;
+        }
+
+        public void Deserialize(IDictionary<string, object> dictionary, JavaScriptSerializer serializer)
+        {
+            foreach (string key in dictionary.Keys)
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs]: Why it uses foreach + switch? Looks giant/weird just to set some fields from a dictionary. Is it a pattern their established around in their code?

File: source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptioning.cs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        public long? Id { get; set; }
+
+        /// <summary>
+        /// A set of sources which provide caption. 
+        /// Only one CaptionSource is supported at this time.
+        /// </summary>
+        public ICollection<BrightcoveCaptionSource> CaptionSources { get; set; }
+
+        public IDictionary<string, object> Serialize(
+            JavaScriptSerializer serializer)
```
Comment (denmarkin) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptioning.cs]: That's quite a mish mash of different formatting styles. I guess we should stick to their (long oneliners?)

File: tests/BrightcoveOS .NET-MAPI-Wrapper.Tests/IntegrationTests/VideoWrite/AddCaptioningTest.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+using NUnit.Framework;
+
+namespace BrightcoveOS.NET_MAPI_Wrapper.Tests.IntegrationTests.VideoWrite
+{
+    [TestFixture]
+    public class AddCaptioningTest : VideoWriteTestBase
+    {
+        //Sample DFXP Files
+        private readonly string FileToUpload = @"..\\..\\Assets\\Sample_CaptionFile.dfxp";
+        private readonly string ExternalCaptionUrl = "https://s3-us-west-2.amazonaws.com/public-rev/captions/Example.dfxp";
```
Comment (denmarkin) [inline tests/BrightcoveOS .NET-MAPI-Wrapper.Tests/IntegrationTests/VideoWrite/AddCaptioningTest.cs]: fragile, should be hosted by repo owner

**Comment by denmarkin** [comment]
Sorry guys, didn't realize it's the main repo and was commenting in context of our fork.

File: source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs (hunk 3 above)
Comment (amhed) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs]: I left it as their API docs said. We can change it to use what suits us in our code.

File: source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs (hunk 6 above)
Comment (amhed) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs]: Convention

File: source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs (hunk 7 above)
Comment (amhed) [inline source/BrightcoveOS .NET-MAPI-Wrapper/Model/Items/BrightcoveCaptionSource.cs]: You're right, I should update these.

File: tests/BrightcoveOS .NET-MAPI-Wrapper.Tests/IntegrationTests/VideoWrite/AddCaptioningTest.cs (hunk 10 above)
Comment (amhed) [inline tests/BrightcoveOS .NET-MAPI-Wrapper.Tests/IntegrationTests/VideoWrite/AddCaptioningTest.cs]: I'll look for a more generic example

**Comment by dmillz** [comment]
Hey guys, this is fantastic! I love seeing contributions back from the community :-D

Unfortunately, I'm not really set up to work on .NET code these days. I originally wrote this wrapper 2 jobs and over 4 years ago, and have been occassionally maintaining it in in my spare time, pulling in new contributions every once in a while. As you might be able to tell from the other open [6-month-old pull request](https://github.com/BrightcoveOS/.NET-MAPI-Wrapper/pull/35) that I've let languish, my spare time is in short supply these days.

I'm going to ping my friends and former colleagues at Velir to ask if anyone there wants to take over maintenance of this project, since the copyright technically belongs to them. If not Velir, then I'm willing to bet that someone at Brightcove would step up to the plate. 

Thanks, and stay tuned.

**Comment by dmills-yesware** [comment]
FYI,  I reached out to Velir, but no one there has expressed interest in taking over the project. I emailed Brightcove this morning, and they're going ask around to see if anyone might want to take our in. 

The thought occurs to me, though: as active users of the project, would **you** possibly be interested in being added to the project as collaborators? That way you could merge whatever changes you desire and issue official releases.

**Comment by dmills-yesware** [comment]
Oops, this is still the same dmillz, I'm just logged in under my work account :-)

**Comment by amhed** [comment]
Sure, I'm up for updating it (and probably writing another wrapper for their other API, their ingestion engine)

**Comment by dmills-yesware** [comment]
Fantastic! I don't think I actually have permission on this repo to add collaborators, but I'll ask Brightcove to add you. Should we add @denmarkin too?

**Pull request merged**